libyul/backends/evm/StackLayoutGenerator.cpp

```diff
@@ -0,0 +1,560 @@
+			bool nonZeroVisited = _blocksWithExitLayout.count(_conditionalJump.nonZero);
+			if (zeroVisited && nonZeroVisited)
+			{
+				Stack stack = combineStack(
+					m_layout.blockInfos.at(_conditionalJump.zero).entryLayout,
+					m_layout.blockInfos.at(_conditionalJump.nonZero).entryLayout
+				);
+				stack.emplace_back(_conditionalJump.condition);
+				return stack;
+			}
+			if (!zeroVisited)
+				_toVisit.emplace_front(_conditionalJump.zero);
+			if (!nonZeroVisited)
+				_toVisit.emplace_front(_conditionalJump.nonZero);
+			return nullopt;
+		},
+		[&](CFG::BasicBlock::FunctionReturn const& _functionReturn) -> std::optional<Stack>
+		{
+			yulAssert(_functionReturn.info, "");
+			Stack stack = _functionReturn.info->returnVariables | ranges::views::transform([](auto const& _varSlot){
+				return StackSlot{_varSlot};
+			}) | ranges::to<Stack>;
+			stack.emplace_back(FunctionReturnLabelSlot{});
+			return stack;
+		},
+		[&](CFG::BasicBlock::Terminated const&) -> std::optional<Stack>
+		{
+			return Stack{};
+		},
+	}, _block.exit);
+}
+
+void StackLayoutGenerator::processEntryPoint(CFG::BasicBlock const& _entry)
+{
+	list<CFG::BasicBlock const*> toVisit{&_entry};
+	set<CFG::BasicBlock const*> visited;
+
+	// TODO: check whether visiting only a subset of these in the outer iteration below is enough.
+	list<pair<CFG::BasicBlock const*, CFG::BasicBlock const*>> backwardsJumps = collectBackwardsJumps(_entry);
+
+	while (!toVisit.empty())
+	{
+		// First calculate stack layouts without walking backwards jumps, i.e. assuming the current preliminary
+		// entry layout of the backwards jump target as the initial exit layout of the backwards-jumping block.
+		while (!toVisit.empty())
+		{
+			CFG::BasicBlock const *block = *toVisit.begin();
+			toVisit.pop_front();
+
+			if (visited.count(block))
+				continue;
+
+			if (std::optional<Stack> exitLayout = getExitLayoutOrStageDependencies(*block, visited, toVisit))
+			{
+				visited.emplace(block);
+				// We can skip the visit, if we have seen this precise exit layout already last time.
+				// Note: if the entire graph is revisited in the backwards jump check below, doing
+				//       this seems to break things; not sure why.
+				// Note: since I don't quite understand why doing this can break things, I comment
+				//       it out for now, since not aborting in those cases should always be safe.
+				// if (auto* previousInfo = util::valueOrNullptr(m_layout.blockInfos, block))
+				//	if (previousInfo->exitLayout == *exitLayout)
+				//		continue;
+				auto& info = m_layout.blockInfos[block];
+				info.exitLayout = *exitLayout;
+				info.entryLayout = propagateStackThroughBlock(info.exitLayout, *block);
+
+				for (auto entry: block->entries)
+					toVisit.emplace_back(entry);
+			}
+			else
+				continue;
+		}
+
+		// Determine which backwards jumps still require fixing and stage revisits of appropriate nodes.
+		for (auto [block, target]: backwardsJumps)
+			if (ranges::any_of(
```

Comment:
Can you summarize the condition? There is a slot in the exit layout of the block that jumps that is not in the entry layout of the block we are jumping to? Can this be done using `std::set` intersection somehow? But maybe a comment is enough.

Reply:
Oh, the explanation is below :)

Reply:
Yeah, the condition is basically: at the time of backwards jumping I'm missing a slot that the jump target needs, i.e. I can't jump back to the loop condition, since I don't have all "input slots" of the loop at hand anymore. (The problem is that I have to start *somewhere* to figure out what the loop condtiion block actually needs, so the first iteration just assumes that the backwards jump needs nothing and - in the best case of a non-nested loop - the second iteration will know what variables I need to keep through the loop).

I'll try to add a good comment.

Reply:
just pull the first lines of the one below up